src/ai-jsx/util.ts

```diff
@@ -0,0 +1,372 @@
+      }
+
+      // Stop if we encounter another chat component
+      if (child.name === makeComponentName(OpenAIChatModel)) {
+        continue;
+      }
+
+      walk(child);
+    }
+  };
+
+  walk(completionSpan);
+
+  return tracking;
+}
+
+export async function processCompletedRootSpan(rootSpan: ABSpan) {
+  const traceId = rootSpan.id;
+  const events: { message: string; args: SendEventArgs }[] = [];
+  const seenTrackerIds = new Set<string>();
+
+  const walk = (span: ABSpan, parentCompletionSpanId?: string) => {
+    let completionSpanId: string | undefined = parentCompletionSpanId;
+
+    if (span.name === makeComponentName(OpenAIChatModel)) {
+      completionSpanId = span.id;
+      const trackerId = parseTrackerIdFromProps(span.props);
+
+      if (!trackerId || !seenTrackerIds.has(trackerId)) {
+        // TODO: Need to figure out how to differentiate between which messages were
+        // sent in the request and which were part of the response, since they're all
+        // rendered together as children under <ChatCompletion>. For now just assuming
+        // that all system and user messages are part of the request and all assistant
+        // messages are part of the response.
+        const requestMessages = makeMessagesArrayFrom(span, [
+          SystemMessage,
+          UserMessage,
+        ]);
+        const responseMessages = makeMessagesArrayFrom(span, [
+          AssistantMessage,
+        ]);
+
+        events.push({
+          message: 'ai.completion.request',
+          args: {
+            traceId,
+            spanId: completionSpanId,
+            parentSpanId: parentCompletionSpanId,
+            timestamp: span.startTime,
+            properties: {
+              ...Object.fromEntries(
+                Object.entries(span.props as Record<string, unknown>).filter(
+                  ([k]) => k !== AUTOBLOCKS_TRACKER_ID_PROP_NAME,
+                ),
+              ),
+              messages: requestMessages.map((m) => ({
+                role: m.role,
+                content: m.content,
+              })),
+              provider: 'openai',
+            },
+          },
+        });
+
+        if (span.endTime) {
+          events.push({
+            message: 'ai.completion.response',
+            args: {
+              traceId,
+              spanId: completionSpanId,
+              parentSpanId: parentCompletionSpanId,
+              timestamp: span.endTime,
+              properties: {
+                latency:
+                  new Date(span.endTime).getTime() -
+                  new Date(span.startTime).getTime(),
+                choices: responseMessages.map((m) => ({
+                  message: { role: m.role, content: m.content },
+                })),
+                usage: {
+                  completion_tokens: responseMessages.reduce(
+                    (acc, m) => acc + (m.tokens || 0),
+                    0,
+                  ),
+                },
+                provider: 'openai',
```

Comment:
does ai jsx only work with openai?
Do we have model information here too?

Reply:
and yeah model info comes from the ChatCompletion props

Reply:
Can followup but we should determine provider from model in that case